include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
   }
 
   bool hasPredecessors() const { return Predecessors.size() != 0; }
+  size_t predecessorCount() const { return Predecessors.size(); }
+
+  typename Base::DerivedType *&predecessorAt(size_t Index) {
```

Comment:
Yeah, it was intentional. I was thinking about giving an access for pointer modification, but now that I think about how important the pointer stability is, it was a bad idea. Thanks for pointing this out.

Reply:
Wait-wait-wait-wait, I misunderstood where this code was.
This part (a part of the `BidirectionalNode`) does actually allow for explicit edge modification, mostly, this return type mimics the return type Massimo (thanks, git blame) chose for the same thing (see `ForwardNode::getNeighbor`).

Reply:
On the second note, searching through the sources for `Base::DerivedType` gave exactly two results, and they are both here. So, what do you mean by "quite often"?

Reply:
> Wait-wait-wait-wait, I misunderstood where this code was.
> This part (a part of the `BidirectionalNode`) does actually allow for explicit edge modification, mostly, this return type mimics the return type Massimo (thanks, git blame) chose for the same thing (see `ForwardNode::getNeighbor`).

You can still change the edge if you return a pointer. But returning a *reference* to the pointer. What's the use case for that? Maybe it makes sense if it's a `const` reference. But a non-`const` is smelly.

> On the second note, searching through the sources for Base::DerivedType gave exactly two results, and they are both here. So, what do you mean by "quite often"?

Uhm, OK. Up to you, I just dislike `typename` in function prototypes :)

Reply:
> You can still change the edge if you return a pointer. But returning a reference to the pointer. What's the use case for that? Maybe it makes sense if it's a const reference. But a non-const is smelly.

What this is returning is not an edge, but the neighbor node which is stored as a pointer (!). So we're returning the reference to a variable we're storing internally (which just happens to have a pointer type).

The purpose is exactly as it sounds, to let the user change the pointer itself (for example, to make this edge point to a different node without changing the edge itself).

Reply:
> Uhm, OK. Up to you, I just dislike typename in function prototypes :)
So do I, but for me it's mostly "if it fits into a single line, I can live with it.

I don't see much reason to turn it into a three shorter lines rather than two longer ones. I guess, it's more of a personal preference.

Reply:
Fair enough.